### PR TITLE
Fix `big(typemin(T))`

### PR DESCRIPTION
**Pull request by martinholters**
master:
```julia
julia> big(typemin(Int8))
-18446744073709551488

julia> big(typemin(Int16))
-18446744073709518848
```
this PR:
```julia
julia> big(typemin(Int8))
-128

julia> big(typemin(Int16))
-32768
```
(Bug was introduced by #29165, i.e. after 1.0.)

### CODE REVIEW

**Comment by martinholters** [comment]
"Before" = master
"After" = this PR
(Clarified in OP)
Does that answer your question, or is the question indeed what's wrong with the behavior of master?

File: base/gmp.jl [hunk 1]
```diff
@@ -287,7 +287,11 @@ function BigInt(x::Integer)
     nd = ndigits(x, base=2)
     z = MPZ.realloc2(nd)
     s = sign(x)
-    s == -1 && (x = -x)
+    if x isa Union{Int8,Int16,Int32,Int64,Int128} # two's complement
+        x = s == -1 ? Unsigned(~x) + one(x) : Unsigned(x)
+    else
+        s == -1 && (x = -x)
```
[the remark this answers is not in the document]
Comment (rfourquet) [reply base/gmp.jl]: I think we need to have an abstract type between `Signed` and those builtin types to represent this class of types, e.g. `SignedTwoComplement` or something (or a trait indicating that `Unsigned` can be applied to the number). For now, what about a mix of your solution and your idea to check that `x > 0`:
```julia
s == -1 && (x = -x)
if x < 0
    x = s == -1 ? Unsigned(~x) + one(x) : Unsigned(x)
end
```
This may throw at the `Unsigned` call for number which don't support that, but only if the basic `x = -x` already fails.
Would there be a type-instability performance problem due to having `x <: Unsigned` based on a dynamic condition?

[the remark this answers is not in the document]
Comment (martinholters) [reply base/gmp.jl]: Yes, that would introduce type-instability, although the performance-penalty may be acceptably low with the `Union` optimizations. However, just the fact that `x < 0 && -x < 0` does not necessarily mean that `x` is in two complement (although it is _very_ likely, of course.)

I've also considered introducing an `unsigned_abs` that returns the absolute value as an `Unsigned` type (or rather, a type which can definitely hold `abs(x)` and may be unsigned) which could then be implemented for new types as needed. Introducing `SignedTwoComplement` sounds more attractive to me, but adding a new abstract type in the `Number`hierarchy is, of course, a change that should be well thought through.

That said, the present code is simple and a strict improvement :smile:

**Comment by oscardssmith** [comment]
Yep, makes sense. I got the order backwards.

File: base/gmp.jl (hunk 1 above)
[the remark this answers is not in the document]
Comment (martinholters) [reply base/gmp.jl]: Ah, but that should be `unsigned` rather than `Unsigned` because the latter unnecessarily checks that its argument is not negative which we already know here. Will update locally and push once we agree on the general path forward here.

[the remark this answers is not in the document]
Comment (rfourquet) [reply base/gmp.jl]: But doens't that assume anyway two's complement? the previous code (before the bug) was doing something like that, taking the bit pattern of `x < 0` and at the end returning `-b-1` where `b` was the constructed `BigInt`. So I actually think we can assume here two's complement without breaking code. 
One of my motivation doing this change was to accelerate converting bigger primitive types, like `Int512`, so it would be great if the fix works for non-base types!
Don't you think that if `x` is transformed to `Unsigned` in the case `x < 0 && -x < 0`, then the type-instability possible performance hit would occur only for this rare corner case? This seems simpler than implementing the "carry carried across the loop".

File: test/bigint.jl [hunk 2]
```diff
@@ -432,3 +432,8 @@ end
     y = rand(UInt128)
     @test BigInt(y) % UInt128 === y
 end
+
+@testset "conversion from typemin($T)" for T in (Int8, Int16, Int32, Int64, Int128)
+    x = typemin(T)
+    @test big(x) == x
```
Comment (simonbyrne) [inline test/bigint.jl]: Is this actually that useful of a test? If the comparison uses promotion, then it won't work correctly.

Comment (martinholters) [reply test/bigint.jl]: True. Currently, comparison does not promote (that's how I stumbled upon this), but if it does in the future, this test will become useless. I'll change to use explicit constants.

**Comment by martinholters** [comment]
Ok, having slept this over, KISS seems best here.

I'm not even sure we can truely support general, non-two complement integer representations. Apart from the negation I deal with here, we also use `>>>`, which is defined as a bit shift, which may not make any sense for arbitrary integer representations. If we rewrite that as `÷ 2^n`, we assume the division to be well-behaved (for all inputs), which is likely, but in no way guaranteed. I think we would need a fused `(x ÷ 2^n) % T`operation, which (for suitable type `T`) does the right thing and can be implemented by custom integer types. I doubt it's worth it and definitely beyond the scope of this PR which is intended to be a small bugfix.

**Review by rfourquet — APPROVED** [review]
Thanks for this beautiful KISS solution!
